File: Other Docs/geocoding.R
# --------------------------------------------------------------------------------------------------------------
# --------------------------------------------------------------------------------------------------------------
#                                                 GeoCoding
# --------------------------------------------------------------------------------------------------------------
#---------------------------------------------------------------------------------------------------------------

library(geosphere)
library(RJSONIO)
# API to return details about the address
googleGeocodeUrl <- 'http://maps.google.com/maps/api/geocode/xml?sensor=false&address='

# --------------------------------------------------------------------------------------------------------------
#                         Function to call the API and retrieve lat and long from response  
# --------------------------------------------------------------------------------------------------------------

getDocNodeVal=function(doc, path)
{
  sapply(getNodeSet(doc, path), function(el) xmlValue(el))
}

gGeoCode=function(str)
{
  library(XML)
  u=paste(googleGeocodeUrl,str)
  doc = xmlTreeParse(u, useInternal=TRUE)
  str=gsub(' ','%20',str)
  lat=getDocNodeVal(doc, "/GeocodeResponse/result/geometry/location/lat")
  lng=getDocNodeVal(doc, "/GeocodeResponse/result/geometry/location/lng")
  list(lat = lat, lng = lng)
}

LAT_LONG_1 <- gGeoCode("majestic, Bangalore")
LAT_LONG_2 <- gGeoCode("manyata tech park, Bangalore")

# Reference : http://www.r-bloggers.com/calling-google-maps-api-from-r/
#             https://cran.r-project.org/web/packages/geosphere/geosphere.pdf

# --------------------------------------------------------------------------------------------------------------
#             Function that contain the algorithm to calculate the displacement between two co-ordinates
# --------------------------------------------------------------------------------------------------------------

distm(as.numeric(paste(LAT_LONG_1)),as.numeric(paste(LAT_LONG_2)),fun=distHaversine)

                                          #   OR   #  

distHaversine(as.numeric(paste(LAT_LONG_1)), as.numeric(paste(LAT_LONG_2)), r=6378137)

# --------------------------------------------------------------------------------------------------------------
# --------------------------------------------------------------------------------------------------------------
# ------------------- END --------------------- GeoCoding --------------------------- END ----------------------
# --------------------------------------------------------------------------------------------------------------
#---------------------------------------------------------------------------------------------------------------

File: Other Docs/geocoding.py
import urllib
import json

    #API to return details about the address
googleGeocodeUrl = 'http://maps.google.com/maps/api/geocode/xml?sensor=false&address='

# Function to call the API and retrieve lat and long from response  
def get_coordinates(query, from_sensor=False):
    query = query.encode('utf-8')
    #Parameters to send while calling the API
    params = {
        'address': query,
        'sensor': "true" if from_sensor else "false"
    }
    #Create the API and Call the API
    url = googleGeocodeUrl + urllib.urlencode(params)
    json_response = urllib.urlopen(url)
    response = json.loads(json_response.read())
    #Check if results are present in the response and extract the lat and long from it
    if response['results']:
        location = response['results'][0]['geometry']['location']
        latitude, longitude = location['lat'], location['lng']
        print query, latitude, longitude
    else:
        latitude, longitude = 0.0, 0.0
        print query, "<no results>"
    return latitude, longitude

#Function that contain the algorithm to calculate the displacement between two co-ordinates
from math import radians, cos, sin, asin, sqrt
def haversine(lon1, lat1, lon2, lat2):
    # convert decimal degrees to radians 
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
    # haversine formula 
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    km = 6367 * c
    return km

#Input address from user
print 'Enter the address'
add = raw_input()
lat,lon = get_coordinates(add)
